src/client/metric.rs

```diff
@@ -286,11 +291,170 @@ impl<T: MetricType + Clone> Metric<T> {
         self.mmap_view = mmap_view;
     }
 
-    pub (super) fn write_val(&self, cursor: &mut Cursor<&mut [u8]>) -> io::Result<()> {
-        self.val.write_to_writer(cursor)
+    pub (super) fn write_val<W: Write>(&self, writer: &mut W) -> io::Result<()> {
+        self.val.write_to_writer(writer)
     }
 }
 
+pub struct InstanceMetric<T> {
+    pub (super) metrics: HashMap<String, Metric<T>>,
```

Comment:
I see that your idea of an instance metric is multiple Singleton Metrics wrapped inside a Combined metric type with instances, but doesn't that seem an overkill, especially since all other attributes (semantics, unit, help) will be the same for all internal metrics. Maybe we can have a `metricData` type managing singleton data used by both Singleton and Instance Metrics.